sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamEnumerableConverter.java

```diff
@@ -201,7 +204,9 @@ private static void runCollector(PipelineOptions options, BeamRelNode node) {
     PCollection<Row> resultCollection = BeamSqlRelUtils.toPCollection(pipeline, node);
     resultCollection.apply(ParDo.of(new Collector()));
     PipelineResult result = pipeline.run();
-    result.waitUntilFinish();
+    if (PipelineResult.State.FAILED.equals(result.waitUntilFinish())) {
+      throw new RuntimeException("Pipeline failed for unknown reason");
```

Comment:
Noting that for the DirectRunner, we want `result.waitUntilFinish()` to always throw an exception when there's a failure. I think it is always a bug if you get FAILED status without throwing.

Reply:
You may want to file a bug. The behavior before this PR is that an exception will be thrown `at most once`, after the PR it is `exactly once`. If there are multiple threads waiting on `result.waitUntilFinish()`, not all will receive exceptions.